### PR TITLE
Timestamp kit backward compatibility update

### PR DESCRIPTION
**Pull request by jasoncypret**
#### Screens

This PR was originally revered as it wasn't backward compatible and caused breaking changes on Nitro.

## Rails
Screen Shot 2020-12-21 at 10 24 14 AM
Screen Shot 2020-12-21 at 10 24 26 AM

## React
Screen Shot 2020-12-21 at 10 24 40 AM
Screen Shot 2020-12-21 at 10 24 48 AM

#### Runway Ticket URL

Github Issue: [https://github.com/powerhome/playbook/issues/618](https://github.com/powerhome/playbook/issues/618)

#### How to test this

React Kit should be accessible at: /kits/timestamp/react
Rails Kit at: /kits/timestamp/rails

#### Checklist:

- [x] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [x] **URGENCY** Please select a release milestone
- [x] **DEPLOY** Please add the `Milano` label when you are ready for a review.
- [x] **SCREENSHOT** Please add a screen shot or two.
- [x] **SPECS** Please cover your changes with specs.
- [x] **READ DOCS** Please make sure you have read and understand the [Playbook Release Process](https://github.com/powerhome/playbook/wiki/Playbook-Releases)

*The normal release cut off deadline is 3p EDT each week. Please reach out to the release team if you have an urgent request or need a release off cycle.*

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb [hunk 1]
```diff
@@ -1,6 +1,16 @@
...
+  aria: object.aria,
+  id: object.id,
+  data: object.data,
+  class: object.classname) do %>
+
+  <% if object.variant == "updated" || object.variant == "elapsed" %>
+    <%= pb_rails("caption", props: { text: object.format_updated_string, size: 'xs', dark: object.dark }) %>
+  <% else %>
+    <% if object.show_date %>
+      <%= pb_rails("caption", props: { text: object.timestamp ? object.format_datetime_string : object.text, size: 'xs', dark: object.dark }) %>
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb:11]: Ternaries should probably be abstracted to a method. Think of a method name that suits this logic and move this part.

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
+    <% else %>
+      <%= pb_rails("caption", props: { text: object.format_time_string, size: 'xs', dark: object.dark }) %>
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb:13]: What is interesting here is that both `size` and `dark` props are the same values while only the `text` prop changes based on the outer conditional. It is entirely possible to dry this part of a bit. Parsing through a lot of conditional logic in erb is feeling a little damp.

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx [hunk 3]
```diff
@@ -2,36 +2,116 @@
...
+    if (dateTimestamp.toYear() > currentYear) {
+      fullDisplay = `${fullDisplay}, ${dateTimestamp.toYear()}`
+    }
+    return `${fullDisplay} ${' \u00b7 '} ${fullTimeDisplay(dateTimestamp, timeDisplay, timezone, showTimezone)}`
+  }
+
+  const formattedUpdatedString = (showUser, text, dateTimestamp, timeDisplay) => {
+    switch (variant) {
+    case 'elapsed': {
+      const elapsedUserDisplay = (showUser == true && text.length > 0) ? ` by ${text}` : ''
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx:76]: Similar to the comment above in the template, it makes sense to break this down a little more for readability here. 

```suggestion
      const shouldShowUser = showUser == true && text.length > 0
      const elapsedUserDisplay = shouldShowUser ? ` by ${text}` : ''
```

Now it will be easier to re-visit this area and edit the string interpolation area without having to visually parse through the ternary logic comparator as well.

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx [hunk 4]
```diff
@@ -2,36 +2,116 @@
...
+    case 'updated': {
+      const updatedUserDisplay = (showUser == true && text.length > 0) ? ` by ${text}` : ''
+      return `Last updated ${updatedUserDisplay} on ${dateDisplay} at ${timeDisplay}`
+    }
+    }
+  }
 
   return (
     <div
+        {...ariaProps}
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx:88]: All spreads must be *after* regular props. I realize this was done wrong originally, but it would be the time to correct that.

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx [hunk 5]
```diff
@@ -2,36 +2,116 @@
...
+                text={timestamp ? fullDateDisplay(dateTimestamp, currentYear, dateDisplay, timezone, showTimezone) : text}
+            />
+            <Else />
+            <Caption
+                dark={dark}
+                size="xs"
+                text={fullTimeDisplay(dateTimestamp, timeDisplay, timezone, showTimezone)}
+            />
+          </If>
+        </If>
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx:113]: Similar to the above comment on the Rails `.erb` template, you can apply the same thinking here and abstract in a very similar way in order to dry this up a bit.

File: playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx [hunk 6]
```diff
@@ -5,42 +5,18 @@ const TimestampUpdated = (props) => {
...
-      <br />
-      <br />
-
-      <Timestamp
-          showTimezone="true"
-          showUser="true"
-          text="Maricris Nonato"
-          timestamp={new Date().getTime()}
-          timezone="America/New_York"
+          timestamp={new Date((new Date()).getFullYear() + 4, (new Date().getMonth()), (new Date().getDate()), (new Date().getHours()), (new Date().getMinutes()))}
```
Comment (thestephenmarshall) [inline playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx:10]: This might be a little hard to follow for a doc example. I am having trouble parsing it.

**Review by thestephenmarshall — CHANGES_REQUESTED** [review]
Great start! I left some comments on some optimizations we can try out here. Also, I think we need to get some JS unit testing in place for this one so that we can truly say it is "backwards compatible" with any level of confidence on the React side. If you want, I can add a commit that gives that baseline unit test then you can update it based on your changes.

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb (hunk 2 above)
Comment (stephenmckeon) [reply playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.html.erb:13]: I was thinking the same thing actually! I'll work on that

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx (hunk 4 above)
Comment (stephenmckeon) [reply playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx:88]: @thestephenmarshall how would I go about correcting that then?

File: playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx (hunk 5 above)
Comment (stephenmckeon) [reply playbook/app/pb_kits/playbook/pb_timestamp/_timestamp.jsx:113]: Agreed!

File: playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx (hunk 6 above)
Comment (stephenmckeon) [reply playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx:10]: Understandable for sure. I had a hard figuring it out in the first place. Not sure how else to go about it though if we want to display the full functionality of the kit. Any ideas? @thestephenmarshall

**Comment by stephenmckeon** [comment]
@thestephenmarshall  Besides the existing tests failing, what do you think about the updates? Views are much cleaner now and much more readable. Just have the two questions on your comments above and if you could add that JS unit tests like you mentioned that would be a huge help. thanks!

File: playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx (hunk 6 above)
Comment (thestephenmarshall) [reply playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx:10]: Abstract this inline code to variables would be a great start. Make sure to use meaningful names for the variables so that the doc is sort of self-explanatory.

**Review by thestephenmarshall — APPROVED** [review]
Looks better. All I see that needs updating is `playbook/app/pb_kits/playbook/pb_timestamp/docs/_timestamp_updated.jsx` which is just moving the inline code to meaningful variables.

**Review by bh247484 — APPROVED** [review]
